src/base/static/components/organisms/map-measurement-overlay.tsx

```diff
@@ -0,0 +1,467 @@
+  <span
+    onClick={props.onClick}
+    css={css`
+      border: 2px solid transparent;
+      border-radius: 4px;
+      padding: 4px 4px 0 4px;
+      margin-right: 8px;
+      background-color: ${props.isSelected ? "#999" : "initial"};
+
+      &:hover {
+        border-color: ${props.isEnabled ? "#999" : "transparent"};
+        cursor: ${props.isEnabled ? "pointer" : "unset"};
+      }
+    `}
+  >
+    {props.children}
+  </span>
+);
+
+//const MeasurementToolIcon = styled("span")(
+//  (props: MeasurementToolIconProps) => ({
+//    border: "2px solid transparent",
+//    borderRadius: "4px",
+//    padding: "4px 4px 0 4px",
+//    marginRight: "8px",
+//    backgroundColor: props.isSelected ? "#999" : "initial",
+
+//    "&:hover": {
+//      borderColor: props.isEnabled ? "#999" : "transparent",
+//      cursor: props.isEnabled ? "pointer" : "unset",
+//    },
+//  }),
+//);
+
+MeasurementToolIcon.defaultProps = {
+  isEnabled: true,
+};
+
+const FEET_PER_MILE = 5280;
+const SQUARE_METERS_PER_ACRE = 0.000247105;
+const measurementFormatter = new Intl.NumberFormat("en-US");
+const MIN_POSITIONS = {
+  "create-polygon": 4, // Including final position matching the first.
+  "create-polyline": 2,
+};
+
+const getNewViewport = ({
+  latitude,
+  longitude,
+  zoom,
+  pitch,
+  bearing,
+  altitude,
+  width,
+  height,
+}: WebMercatorViewportOptions): IWebMercatorViewport =>
+  new WebMercatorViewport({
+    latitude,
+    longitude,
+    zoom,
+    pitch,
+    bearing,
+    altitude,
+    width,
+    height,
+  });
+
+const buildMeasurementFeatureCollection = (
+  selectedTool: string,
+  positions: number[][],
+): FeatureCollection => {
+  let featureFn;
+  let newPositions;
+  let numPositions = 0;
+  switch (selectedTool) {
+    case "create-polygon":
+      featureFn = polygon;
+      // Ensure the last position of the Polygon matches the first.
+      newPositions = [positions.concat([positions[0]])];
+      numPositions = newPositions[0].length;
+      break;
+    case "create-polyline":
+      featureFn = lineString;
+      newPositions = positions;
+      numPositions = newPositions.length;
+      break;
+    default:
+      featureFn = () => [];
+      newPositions = positions;
+      // eslint-disable-next-line no-console
+      console.error(`Measurement overlay: unsupported tool ${selectedTool}`);
+      break;
+  }
+
+  return featureCollection(
+    positions.map(position => point(position)).concat(
+      // Only add the measurement feature if we have enough positions on
+      // the map to support the given feature type.
+      numPositions >= MIN_POSITIONS[selectedTool]
+        ? featureFn(newPositions)
+        : [],
+    ),
+  );
+};
+
+const redraw = ({
+  project,
+  ctx,
+  measurementFeatureCollection,
+  width,
+  height,
+}) => {
+  function projectPoint(lon, lat) {
+    const point = project([lon, lat]);
+    // @ts-ignore
+    this.stream.point(point[0], point[1]);
```

Comment:
Not sure how to type this...

Reply:
I guess the issue here was that the `this` binding is caller-specific, designed to run as a callback inside the `d3-geo` package. You can pass a fake `this` parameter to the callback to solve the typing issue, though I'm not sure how useful it is...

Reply:
You can also type cast the `this` variable, which I think makes it more clear that we had a fight with the compiler and want it to "just work":

```suggestion
    (this as any).stream.point(point[0], point[1]);
```

I'd prefer this over having the compiler ignore the entire line, or passing in fake params, etc.

Reply:
I think in general, type casting is a bit dirty and a sign of sloppy typing, so we should use it sparingly.

Reply:
`(this as any).stream.point(point[0], point[1]);` still results in a compiler error:

```
'this' implicitly has type 'any' because it does not have a type annotation.
```

I couldn't find an exact typing for the `this` context here in the `@types/d3-geo` package, though I could be missing something. I made a custom typing that I think describes the `this` context in which the callback runs.